File: PagingDemo/app/src/main/java/com/example/pagingdemo/RecyclerViewAdapter.java
package com.example.pagingdemo;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

public class RecyclerViewAdapter extends PagedListAdapter<User, RecyclerViewAdapter.MyViewHolder> {


    protected RecyclerViewAdapter() {
        super(new DiffUtil.ItemCallback<User>() {
            @Override
            public boolean areItemsTheSame(@NonNull User oldItem, @NonNull User newItem) {
                return oldItem.getId() == newItem.getId();
            }

            @Override
            public boolean areContentsTheSame(@NonNull User oldItem, @NonNull User newItem) {
                return oldItem.getName().equals(newItem.getName());
            }
        });
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.rv_item, parent, false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        User user = getItem(position);
        if (user == null) {
            holder.textView.setText("loading ... ...");
        } else {
            holder.textView.setText(user.getName());
        }
    }

    class MyViewHolder extends RecyclerView.ViewHolder {

        TextView textView;

        MyViewHolder(@NonNull View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.textView);
        }
    }
}

File: RoomDemo/app/src/main/java/com/example/roomdemo/RecyclerViewAdapter.kt
package com.example.roomdemo

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Switch
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

class RecyclerViewAdapter(isCardView: Boolean) : ListAdapter<User, RecyclerViewAdapter.MyViewHolder>(DIFF_CALLBACK) {

    private var isCardView = false

    companion object {
        val DIFF_CALLBACK = object: DiffUtil.ItemCallback<User>() {
            override fun areItemsTheSame(
                    oldUser: User, newUser: User): Boolean {
                return oldUser.id == newUser.id
            }

            override fun areContentsTheSame(
                    oldUser: User, newUser: User): Boolean {
                return oldUser.name.equals(newUser.name)
            }
        }
    }
    init {
        this.isCardView = isCardView

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {

        val layoutInflater = LayoutInflater.from(parent.context)
        var view: View? = null
        if (!isCardView) {
            view = layoutInflater.inflate(R.layout.cell_normal, parent, false)
        } else {
            view = layoutInflater.inflate(R.layout.cell_card, parent, false)
        }
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {

        val user = getItem(position)
        holder.numberView?.text = (position + 1).toString()
        holder.nameView?.text = user.name
    }


    class MyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var numberView: TextView? = null;
        var nameView: TextView? = null
        var switch: Switch? = null
        init {
            numberView = itemView.findViewById(R.id.numView)
            nameView = itemView.findViewById(R.id.nameView)
            switch = itemView.findViewById(R.id.switch1)
        }
    }
}
